### PR TITLE
Core/Entities: Handle partial state change for pets

### PR DESCRIPTION
**Pull request by DanVS**
**Changes proposed:**
Some spells like [Feral Spirit](https://dbwotlk.com/?spell=51533) (51533) summon a set of pets under player control. But after any pet of the set dies, player looses control on remaining.

PR enables `PetHandler` to iterate through all of the player controlled pets instead of getting just one. Therefore, partial set of pets changes are processed. List is made by `Unit::ControlList` iterating and saved in `std::vector` container to be used in pet `PetSetAction` and `PetSpellAutocastOpcode` handlers.

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Tests performed:** Does build and works in-game.

### CODE REVIEW

**Comment by illfated** [comment]
Rare to see a PR starting off with everything correct test-wise in its first commit. Good job, I say. 👍

**Comment by Shauren** [comment]
I don't know how is this supposed to behave tbh - in legion feral spirit does not get a pet bar (no multi-summon does)

File: src/server/game/Handlers/PetHandler.cpp [hunk 1]
```diff
@@ -525,66 +525,76 @@ void WorldSession::HandlePetSetAction(WorldPacket& recvData)
         }
     }
 
-    // check swap (at command->spell swap client remove spell first in another packet, so check only command move correctness)
-    if (move_command)
+    std::vector<Unit*> pets;
+    for (Unit::ControlList::iterator itr = GetPlayer()->m_Controlled.begin(); itr != GetPlayer()->m_Controlled.end(); ++itr)
```
Comment (Shauren) [inline src/server/game/Handlers/PetHandler.cpp]: `for (Unit* controlled : _player->m_Controlled)`

File: src/server/game/Handlers/PetHandler.cpp [hunk 2]
```diff
@@ -525,66 +525,76 @@ void WorldSession::HandlePetSetAction(WorldPacket& recvData)
...
+        if ((*itr)->GetEntry() == pet->GetEntry() && (*itr)->IsAlive())
+            pets.push_back(*itr);
+
+    for (std::vector<Unit*>::iterator itr = pets.begin(); itr != pets.end(); ++itr)
```
Comment (Shauren) [inline src/server/game/Handlers/PetHandler.cpp]: `for (Unit* pet : pets)`

File: src/server/game/Handlers/PetHandler.cpp [hunk 3]
```diff
@@ -738,12 +748,22 @@ void WorldSession::HandlePetSpellAutocastOpcode(WorldPacket& recvPacket)
...
-        ((Pet*)pet)->ToggleAutocast(spellInfo, state != 0);
-    else
-        pet->GetCharmInfo()->ToggleCreatureAutocast(spellInfo, state != 0);
+    std::vector<Unit*> pets;
+    for (Unit::ControlList::iterator itr = GetPlayer()->m_Controlled.begin(); itr != GetPlayer()->m_Controlled.end(); ++itr)
+        if ((*itr)->GetEntry() == pet->GetEntry() && (*itr)->IsAlive())
+            pets.push_back(*itr);
 
-    charmInfo->SetSpellAutocast(spellInfo, state != 0);
+    for (std::vector<Unit*>::iterator itr = pets.begin(); itr != pets.end(); ++itr)
```
Comment (Shauren) [inline src/server/game/Handlers/PetHandler.cpp]: `for (Unit* pet : pets)`

File: src/server/game/Handlers/PetHandler.cpp [hunk 4]
```diff
@@ -738,12 +748,22 @@ void WorldSession::HandlePetSpellAutocastOpcode(WorldPacket& recvPacket)
         return;
     }
 
-    if (pet->IsPet())
-        ((Pet*)pet)->ToggleAutocast(spellInfo, state != 0);
-    else
-        pet->GetCharmInfo()->ToggleCreatureAutocast(spellInfo, state != 0);
+    std::vector<Unit*> pets;
+    for (Unit::ControlList::iterator itr = GetPlayer()->m_Controlled.begin(); itr != GetPlayer()->m_Controlled.end(); ++itr)
```
Comment (Shauren) [inline src/server/game/Handlers/PetHandler.cpp]: `for (Unit* controlled : _player->m_Controlled)`

**Review by Shauren — CHANGES_REQUESTED** [review]

**Comment by DanVS** [comment]
Shauren, in 3.3.5 they've action bar. For example, [see this video](https://www.youtube.com/watch?v=fm4mHH5i2Yw).

And switching to ranged `for` breaks this for some reason 0_o (does compile, but do not work anymore). Can you please check this out?

**Comment by Shauren** [comment]
Of course it does not work. If you kill the primary pet (select it from pet frame) then you can no longer command them because client doesn't send any pet control packets at all, this is not something you can fix by modifying said packet handlers (loops work as intended when they are alive)

**Comment by DanVS** [comment]
Well, I made additional testing with this PR, and results are following:

- killing non-primary pet still allows to save control on remaining, so everything is OK there;
- killing primary pet actually makes action bar to **temporally** disappear, but after some time it turns back and you can still control remaining. However, this time is pretty, so most of the summon time they are not responding.

I'd say that PR works 50/50 now. Probably it's not worth to merge?

**Comment by Shauren** [comment]
When I tested it, killing primary pet simply greyed out the pet bar and it remained unusable for the duration of feral spirit

**Comment by DanVS** [comment]
If you press something when bar is greyed it disappear, but then reappear again.

**Comment by jackpoz** [comment]
this is how I handled it 8 years ago: https://github.com/arcemu/arcemu/commit/1e4b99d889314d6578e1f72f33aed6e464990221#diff-12a46c1c88d79d5ab384cc6949e2734eR810

When the main summon died (say wolf A in the Feral Spirit case), I set the main summon to the other summon still alive and sent again the spells to the client.

**Comment by DanVS** [comment]
Jackpoz, thanks! But is seems that this is already handled in [Unit.cpp](https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Entities/Unit/Unit.cpp#L5961-L5992).

**Comment by illfated** [comment]
I presume you mean the 3.3.5 version, although it is identical:
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Unit/Unit.cpp#L5947-L5978
Then again, are we sure that it is handled as intended or the way jackpoz suggested?

**Comment by DanVS** [comment]
I think I figured it out. Action bar (and primary pet) reappear after corpse of died summon had removed.  I can't find a way to change that.

**Comment by DanVS** [comment]
Should be fixed now.

File: src/server/game/Entities/Creature/TemporarySummon.cpp [hunk 5]
```diff
@@ -300,6 +300,21 @@ void Minion::RemoveFromWorld()
     TempSummon::RemoveFromWorld();
 }
 
+void Minion::setDeathState(DeathState s)
+{
+    Creature::setDeathState(s);
+    if (s == JUST_DIED && IsGuardianPet())
```
Comment (Seyden) [inline src/server/game/Entities/Creature/TemporarySummon.cpp]: That just looks so horrible.. couldn't you please do it like:

if (s != JUST_DIED || !IsGuardianPet())
    return;

and so on.. that would look a lot better :D

Comment (Seyden) [reply src/server/game/Entities/Creature/TemporarySummon.cpp]: Smth like that

`void Minion::setDeathState(DeathState s)
{
    Creature::setDeathState(s);
    if (s != JUST_DIED || !IsGuardianPet())
        return;

    Unit* owner = GetOwner();
    if (!owner || !owner->IsPlayer() || owner->GetMinionGUID() != GetGUID())
        return;

    for (Unit* controlled : owner->m_Controlled)
    {
        if (controlled->GetEntry() == GetEntry() && controlled->IsAlive())
        {
            owner->SetMinionGUID(controlled->GetGUID());
            owner->SetPetGUID(controlled->GetGUID());
            owner->ToPlayer()->CharmSpellInitialize();
        }
    }
}
`

**Comment by Killyana** [comment]
DanVS, could test if this issue is fixed or not with your commit https://github.com/TrinityCore/TrinityCore/issues/19770

**Comment by DanVS** [comment]
Is doesn't fix an issue given.

File: src/server/game/Entities/Creature/TemporarySummon.cpp [hunk 6]
```diff
@@ -300,6 +300,27 @@ void Minion::RemoveFromWorld()
...
+    if (!owner || owner->GetTypeId() != TYPEID_PLAYER || owner->GetMinionGUID() != GetGUID())
+        return;
+                
+    for (Unit* controlled : owner->m_Controlled)
+    {
+        if (controlled->GetEntry() == GetEntry() && controlled->IsAlive())
+        {
+            owner->SetMinionGUID(controlled->GetGUID());
+            owner->SetPetGUID(controlled->GetGUID());
+            owner->ToPlayer()->CharmSpellInitialize();
```
Comment (jackpoz) [inline src/server/game/Entities/Creature/TemporarySummon.cpp]: Should there be a break; here ?

Comment (DanVS) [reply src/server/game/Entities/Creature/TemporarySummon.cpp]: It should. Thanks!

File: src/server/game/Entities/Creature/TemporarySummon.cpp (hunk 5 above)
Comment (DanVS) [reply src/server/game/Entities/Creature/TemporarySummon.cpp]: Fixed. Thanks!

**Comment by DanVS** [comment]
@Shauren, @jackpoz is it okay now?

**Comment by jackpoz** [comment]
The little I commented about has been fixed so that's good, I didn't test it ingame tho.

**Comment by illfated** [comment]
In case you were wondering about the Codacy status:
Codacy does not support MySQL yet, so SQL "errors" can be ignored.
The .cpp "error" might be a false positive or not applicable in TC source.

**Comment by DanVS** [comment]
@Shauren, are some changes still requested?

**Comment by jackpoz** [comment]
let's see what happens

**Comment by jackpoz** [comment]
thanks for the PR :)

**Pull request merged**